Makefile: improve some compilation scenario's

maleadt opened the pull request:
This PR contains a bunch of commits which should improve some less-used building scenario's, such as dynamically linking against LLVM, while using CMake, or when using Address Sanitizer, etc. I've also added some comments to document currently-broken scenario's which aren't trivial to fix (or even detect) given the current Makefile infrastructure.

These changes shouldn't break anything, except maybe for dropping `--enable-shared`/`-DLLVM_BUILD_LLVM_DYLIB:BOOL=ON` when configuring LLVM with `LLVM_USE_SHLIB=0`. I did so because building `libLLVM.so` failed in [some circumstances](https://llvm.org/bugs/show_bug.cgi?id=23649); arguably one also shouldn't rely on Julia compiling (but not using) certain libraries unless when specifically asked to do so.

Fix https://github.com/JuliaLang/julia/issues/14921

maleadt committed:
Apply patches before configuring the dependencies.

maleadt committed:
Don't forward sanitize flags to the dependencies.

Some deps don't cope well with ASAN, see #13858.

maleadt committed:
Don't put the CMake buildtype in CMAKE_COMMON.

The LLVM CMake invocation contained the build-type twice.

maleadt committed:
Forcibly disable building of unwanted LLVM components.

maleadt committed:
Fix the LLVM build and check commands.

This now uses "make" instead of "cmake --build" when appropriate,
and fixes fowarding MAKEFLAGS, both now allowing parallel builds.

maleadt committed:
Fix installing LLVM from usr-staging to usr when using CMake.

maleadt committed:
Properly handle USE_LLVM_SHLIB while building.

Also document specific restrictions when doing so in combination with CMake.

maleadt committed:
Pass the correct CMake build type when configuring LLVM.

nalimilan commented on deps/Makefile:
```diff
@@ -441,6 +446,9 @@ endif # ARCH == ppc64
 
 
 ifeq ($(LLVM_SANITIZE),1)
+# NOTE: LLVM_SANITIZE + USE_LLVM_SHLIB only works with LLVM_USE_CMAKE, because
```
Maybe turn this into an actual warning/error when incompatible options are passed?

maleadt replied:
This is a good candidate for an error indeed (I didn't consider it because most other broken scenario's are LLVM version dependent, which is hard to check for).

nalimilan commented on deps/Makefile:
```diff
@@ -485,12 +493,24 @@ LLVM_MFLAGS += OPTIONAL_PARALLEL_DIRS=clang
 else
 # block default building of Clang
 LLVM_MFLAGS += OPTIONAL_PARALLEL_DIRS=
+ifeq ($(LLVM_VER),svn)	# changed since LLVM 3.9
```
If you want to be future-proof, you could also check for `LLVM_VER >= 3.9` (in addition to `svn`).

maleadt replied:
Version checks in Makefiles are reasonably hard to do properly and in a portable manner...

nalimilan replied:
Sigh... Here's an easy way: turn around the `ifeq`, and check whether `LLVM_VER` is in 3.3, 3.4, 3.5, 3.6, 3.7 or 3.8 (you can use `subst` to remove the `.0`, `.1` or `.2` minor release numbers).

maleadt replied:
Why the sigh? I know how to do a plain version comparison, but it's not what I would've liked to upstream.
Anyway, I've updated the PR.

nalimilan replied:
The sigh wasn't against you, it was against the fact that build tools we use in 2016 do not support such elementary operations as version comparisons... :-)

maleadt replied:
Oh yeah, I fully agree with you there! Let's pray for some progress on #11754 :-) (although CMake has its own set of problems too...)

tkelman commented on deps/Makefile:
```diff
@@ -686,29 +706,35 @@ endif
-$(LLVM_OBJ_SOURCE): $(LLVM_BUILDDIR_withtype)/CMakeCache.txt $(LLVM_PATCH_LIST) | $(llvm_python_workaround)
+ifeq ($(CMAKE_GENERATOR),Ninja)
+LLVM_BUILD_COMMAND := $(CMAKE) --build .
+else ifeq ($(CMAKE_GENERATOR),make)
```
there are other generators aside from just these two

maleadt replied:
I based myself on the current handling of `CMAKE_GENERATOR` for LLVM (from `deps/Makefile`):

```make
# For now this is LLVM specific, but I expect it won't be in the future
ifneq ($(LLVM_USE_CMAKE),)
ifeq ($(CMAKE_GENERATOR),Ninja)
CMAKE_GENERATOR_COMMAND := -G Ninja
else ifeq ($(CMAKE_GENERATOR),make)
CMAKE_GENERATOR_COMMAND := -G "Unix Makefiles"
else
$(error Unknown CMake generator '$(CMAKE_GENERATOR)'. Options are 'Ninja' and 'make')
endif
endif
```

maleadt replied:
In the case I added, the fallback scenario could just use `cmake --build`, so I'll fix that. Not sure why the test shown above only allows make or Ninja (@keno?).

tkelman replied:
it was all keno tested

maleadt committed:
Address review comments, and check for LLVM_USE_CMAKE=0|1.

nalimilan commented:
Looks good to me, at least in the parts I understand...

vtjnash commented on deps/Makefile:
```diff
@@ -379,8 +386,8 @@ LLVM_CPPFLAGS := $(CPPFLAGS)
 LLVM_TARGETS := host
 LLVM_TARGET_FLAGS := --enable-targets=$(LLVM_TARGETS)
-LLVM_CMAKE += -DLLVM_TARGETS_TO_BUILD:STRING="$(LLVM_TARGETS)" -DCMAKE_BUILD_TYPE="$(LLVM_BUILDTYPE)" -DLLVM_BUILD_LLVM_DYLIB:BOOL=ON -DLLVM_DYLIB_EXPORT_ALL:BOOL=ON
-LLVM_FLAGS += --disable-profiling --enable-shared --enable-static $(LLVM_TARGET_FLAGS)
```
at some point in the distant past, iirc, the combination of shared and static was required for building llvm on win64 (a032934d). it might be good to check that this doesn't break that target configuration

maleadt replied:
OK, I'll set up a Win64 dev env and see whether nothing broke.

maleadt replied:
Tested on win64 with and without `USE_LLVM_SHLIB`, no regressions.

tkelman commented:
We should also change the llvm svn build to use the git-externals mechanism, though that might be a bigger change than what's here so far.

maleadt committed:
Fix building LLVM with CMake+make on Windows.

For some reason, MSYS2's make doesn't support the recursive
invocation using +$(MAKE).

maleadt commented:
Pushed an additional fix for building LLVM with CMake+make on MSYS2. Should be good to go?

vtjnash commented:
lgtm

maleadt committed:
Properly conditionalize another LLVM config option.

maleadt committed:
Default to LLVM_USE_CMAKE=1 on LLVM 3.9+.

Fixes #14921.

maleadt commented:
Added two commits: I forgot to conditionalize the LLDB option, and a fix for #14921.

maleadt commented:
OK sure, having the default selected based on the LLVM version at hand is useful enough already. I'll force push in a second.

tkelman commented:
are any of the intermediate commits broken here? if not this looks mergeable to me.

maleadt commented:
They should all work; I deliberately kept them separate because Makefile regressions can be annoying to debug.

crayxt commented:
Nice work!
PS: Suddenly cmake 2.8.11 from Centos 7 is too old to build LLVM.

maleadt commented:
Depends on the LLVM version: 3.7 and `trunk` do indeed require CMake 2.8.12.2. But that has always been the case, I didn't change that here.

tkelman commented:
@maleadt did you clear out `usr-staging` before testing this? I'm getting
```
cp: cannot stat ‘/home/tkelman/Julia/julia-0.5/usr-staging/llvm-3.7.1/build_Release/home/tkelman/Julia/julia-0.5/usr/*’: No such file or directory
make[1]: *** [/home/tkelman/Julia/julia-0.5/usr/lib/libLLVMCodeGen.a] Error 1make: *** [julia-deps] Error 2
```

The actual path that does exist is `usr-staging/build/llvm-3.7.1/...`

maleadt commented:
@tkelman Yes, I did multiple `make distcleanall`s as well as fresh checkouts. Can reproduce though, I'll look at it.

crayxt commented:
@maleadt It appears in the buildbot logs http://buildbot.e.ip.saba.us:8010/builders/nightly_llvmsvn-x64

maleadt commented:
OK, found the issue, this should fix it:
```patch
 $(LLVM_OBJ_TARGET): $(LLVM_OBJ_SOURCE) | $(llvm_python_workaround)
 ifeq ($(LLVM_USE_CMAKE),1)
-       $(call staged-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),cd $(LLVM_BUILDDIR_withtype) && $(CMAKE) -DCMAKE_INSTALL_PREFIX="$(call MAKE_DESTDIR,$(LLVM_BUILDDIR_withtype))$(build_prefix)" -P cmake_install.cmake)
+       $(call staged-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),\
+                             cd $(BUILDDIR)/llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE) &&\
+                                                 $(CMAKE) -DCMAKE_INSTALL_PREFIX="$(call MAKE_DESTDIR,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE))$(build_prefix)" -P cmake_install.cmake)
 else
-       $(call make-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),$(LLVM_MFLAGS) PATH="$(llvm_python_workaround):$$PATH")
+       $(call   make-install,llvm-$(LLVM_VER)/build_$(LLVM_BUILDTYPE),\
+                             $(LLVM_MFLAGS) PATH="$(llvm_python_workaround):$$PATH")
 endif # LLVM_USE_CMAKE
        touch -c $@
```
(these Makefiles are getting out of hand)

I don't have time right now, but I'll make a PR later today.

tkelman commented on deps/Makefile:
```diff
@@ -439,8 +463,22 @@ ifeq ($(ARCH), ppc64)
+ifeq ($(USE_LLVM_SHLIB),1)
+ifeq ($(LLVM_USE_CMAKE),1)
+ifeq ($(LLVM_VER_SHORT),$(filter $(LLVM_VER_SHORT),3.3 3.4 3.5 3.6 3.7))
+$(error USE_LLVM_SHLIB=1 with LLVM_USE_CMAKE=1 requires LLVM > 3.7.x)
```
@maleadt which build environments did this error apply to? Could we backport a patch? Then again we might just upgrade to 3.8 very soon after it gets released. I'm testing now to see whether or not this is also an issue on Windows.

maleadt replied:
Only tested on my local linux64 system, where it failed. Looking at the [commit](https://github.com/llvm-mirror/llvm/commit/cd9aae70e988dbd80af933bc144f1c8df431a0e9#diff-54d61d59bfb4db9f734809b8aff1dd11) where it got removed, I'd assume it would be an issue on Windows too.
